### PR TITLE
Corrected condition. Fix for #71596

### LINKED ISSUE #71596: Traffic loss with externalTrafficPolicy:Local and proxy-mode=ipvs
**What happened**:

When a service with `externalTrafficPolicy:Local` is used with proxy-mode=ipvs only 1/nodes traffic gets through. The service have "type: LoadBalancer" and the loadBalancerIP is used for access.

If the service is accessed through the `NodePort` it works as expected.

The problem is that `kube-proxy` setup targets for **all** enpoints in ipvs for the loadBalancerIP while still disable the SNAT. So all traffic that does not happen to hit the own node is lost. Example;
```
> ipvsadm -L -n   # (narrowed)
TCP  192.168.1.4:32713 rr
  -> 11.0.4.3:8080                Masq    1      0          0         
TCP  10.0.0.0:8080 rr
  -> 11.0.1.2:8080                Masq    1      0          0         
  -> 11.0.2.2:8080                Masq    1      0          0         
  -> 11.0.3.2:8080                Masq    1      0          0         
  -> 11.0.4.3:8080                Masq    1      0          0         
```
Here the lbIP `10.0.0.0` gets all endpoints as target (wrong!) but the NotePort entry (`32713`) gets only the local endpoint as target (correct!).

The service manifest;
```
apiVersion: v1
kind: Service
metadata:
  name: mconnect-local
spec:
  selector:
    app: mconnect
  ports:
  - name: mconnect
    port: 5001
  - name: http
    port: 8080
  externalTrafficPolicy: Local
  type: LoadBalancer
```

[metallb](https://github.com/google/metallb) is used to obtain the lbIP.

**What you expected to happen**:

The targets for the `loadBalancerIP` should be the pods executing on the local node **only**, same as for NodePort and no traffic should be lost when using the `loadBalancerIP`.

**How to reproduce it (as minimally and precisely as possible)**:

* Use k8s with proxy-mode=ipvs
* Start a service with `externalTrafficPolicy:Local`
* Try to access the servcie via the loadBalancerIP
* Do `ipvsadm -L -n` on a node and compare entries for loadBalancerIP and NodePort

**Anything else we need to know?**:

Public clouds using an external LB that distributes to the NodePort will not see this bug.

I usually runs on `master` and the problem exist on `` I backed to 1.12 to see if it was a new problem.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.0", GitCommit:"0ed33881dc4355495f623c6f22e7dd0b7632b7c0", GitTreeState:"clean", BuildDate:"2018-09-27T17:05:32Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.0", GitCommit:"0ed33881dc4355495f623c6f22e7dd0b7632b7c0", GitTreeState:"clean", BuildDate:"2018-09-27T16:55:41Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
  No cloud provider. [xcluster](https://github.com/Nordix/xcluster) on Ubuntu 18.04
- OS (e.g. from /etc/os-release):
  Xcluster
- Kernel (e.g. `uname -a`):
  `Linux vm-004 4.18.5 #1 SMP Fri Nov 30 08:47:18 CET 2018 x86_64 GNU/Linux`
- Install tools:
  -
- Others:
  Test using [mconnect](https://github.com/Nordix/mconnect)

/kind bug

### PR DESCRIPTION
**Pull request by uablrek**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

To make `externalTrafficPolicy: Local` work with proxy-mode=ipvs an loadBalancerIP.

**Which issue(s) this PR fixes** 

Fixes #71596 

**Special notes for your reviewer**:

Should be a simple review :-)

**Does this PR introduce a user-facing change?**:

NONE

```release-note
```

/release-note-none

### CODE REVIEW

**Comment by m1093782566** [comment]
/cc @Lion-Wei


--------------------------------------------------
杜军 Du Jun
发件人：k8s-ci-robot
收件人：kubernetes/kubernetes,
抄 送：dujun (D),Review requested,
时间：2018-12-01 01:55:20
主 题：Re: [kubernetes/kubernetes] Corrected condition. Fix for #71596 (#71610)


@k8s-ci-robot<https://github.com/k8s-ci-robot> requested your review on: #71610<https://github.com/kubernetes/kubernetes/pull/71610> Corrected condition. Fix for #71596<https://github.com/kubernetes/kubernetes/issues/71596>.

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -1009,7 +1009,7 @@ func (proxier *Proxier) syncProxyRules() {
 				}
 				if err := proxier.syncService(svcNameString, serv, true); err == nil {
 					// check if service need skip endpoints that not in same host as kube-proxy
-					onlyLocal := svcInfo.SessionAffinityType == v1.ServiceAffinityClientIP && svcInfo.OnlyNodeLocalEndpoints
+					onlyLocal := svcInfo.SessionAffinityType == v1.ServiceAffinityClientIP || svcInfo.OnlyNodeLocalEndpoints
```
Comment (m1093782566) [inline pkg/proxy/ipvs/proxier.go]: @Lion-Wei 

Is this change valid?

**Comment by uablrek** [comment]
/retest

The failed tests are not related to the code change.

**Comment by m1093782566** [comment]
@uablrek 

I think the change breaks the following test case:

```
--- FAIL: TestOnlyLocalLoadBalancing (0.00s)
    proxier_test.go:2797: Expected 1 destinations, got 0 destinations
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 58 [running]:
testing.tRunner.func1(0xc0003de800)
	GOROOT/src/testing/testing.go:792 +0x6a7
panic(0x123a740, 0x1c8fc90)
	GOROOT/src/runtime/panic.go:513 +0x1b9
k8s.io/kubernetes/pkg/proxy/ipvs.checkIPVS(0xc0003de800, 0xc0003c7680, 0xc00038dbc8)
	pkg/proxy/ipvs/proxier_test.go:2800 +0x78a
k8s.io/kubernetes/pkg/proxy/ipvs.TestOnlyLocalLoadBalancing(0xc0003de800)
	pkg/proxy/ipvs/proxier_test.go:1293 +0x42c
testing.tRunner(0xc0003de800, 0x1392e80)
	GOROOT/src/testing/testing.go:827 +0x163
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:878 +0x651
```

**Comment by uablrek** [comment]
I have corrected the test by copying from `TestOnlyLocalNodePorts`. It now passes and also if I revert the fix in this PR it fails with;
```
proxier_test.go:2810: Expected 1 destinations, got 2 destinations
```
as it should.

File: pkg/proxy/ipvs/proxier.go (hunk 1 above)
Comment (lbernail) [reply pkg/proxy/ipvs/proxier.go]: I'm not sure the `SessionAffinityType` test makes sense here:
- when using `externalTrafficPolicy: Local` we should only use local endpoints in any case
- sessionAffinity requires the external LB to ensure affinity (so traffic will only get in the cluster on a given node). If the IPVS service is also configured with persistency we will have end-to-end affinity (and we can load-balance to a pod on a remote host)
- using both at the same time should work (and will ensure sourceIP is preserved)

**Comment by lbernail** [comment]
@uablrek : OK sorry I'm wrong about this: traffic to ExternalIPs on the service port are always load-balanced (not sure why, because I think technically it could work). 

If the service is a NodePort service, accessing it on ExternalIP:Nodeport will not work either because we do not create a VirtualServer for `EIP:NP` (once again, not a technical limitation but I assume iptables works the same)

**Comment by lbernail** [comment]
Ok after checking, if you create a Nodeport service in iptables mode you can access the service on EIP:Nodeport and `externalTrafficPolicy: Local` will work
@uablrek would that help mitigate #62537 for you?

**Comment by m1093782566** [comment]
>  externalIP are not supposed to support `externalTrafficPolicy: Local`

True.

**Comment by lbernail** [comment]
@m1093782566 
> > externalIP are not supposed to support `externalTrafficPolicy: Local`
> True.

Do you know if there is a reason? It looks like a reasonable feature that shouldn't be very hard to implement

**Comment by lbernail** [comment]
To get back to the PR here, I think:
```
  onlyLocal := svcInfo.SessionAffinityType == v1.ServiceAffinityClientIP || 
  svcInfo.OnlyNodeLocalEndpoints
  activeIPVSServices[serv.String()] = true
  activeBindAddrs[serv.Address.String()] = true
  if err := proxier.syncEndpoint(svcName, onlyLocal, serv); err != nil {
```
should be changed to:
```
  activeIPVSServices[serv.String()] = true
  activeBindAddrs[serv.Address.String()] = true
  if err := proxier.syncEndpoint(svcName, svcInfo.OnlyNodeLocalEndpoints, serv); err != nil {
```
because SessionAffinity and ExternalTrafficPolicy: Local are two independent features (unless I missed something)

**Comment by m1093782566** [comment]
> because SessionAffinity and ExternalTrafficPolicy: Local are two independent features (unless I missed something)

AFAIK, it's a work around for LB type service - there is an e2e test though I forget which one. @Lion-Wei should know that.

**Comment by lbernail** [comment]
@m1093782566 I found the PR here: https://github.com/kubernetes/kubernetes/pull/64146

I think the fix for the test is in `syncEndpoint`, the other changes are:
- remove `svcInfo.OnlyNodeLocalEndpoints` for externalIP (related to discussion above)
- test for affinity && OnlyNodeLocalEndpoints for loadbalancerIPs (I don't think we should take affinity into account there)
- remove `svcInfo.OnlyNodeLocalEndpoints` for Nodeport which was wrong and fixed by https://github.com/kubernetes/kubernetes/pull/66064

**Comment by lbernail** [comment]
I can create a quick patch with my suggestion and see if e2e tests are fine (and follow up if there is any issue)

**Comment by m1093782566** [comment]
Yeah,  #64146 is related.

**Comment by lbernail** [comment]
@uablrek I created a new PR with the fix I was suggesting (I included your commits and added mine, but of course I couldn't push to your branch)

**Pull request closed without merging**